plugin/hooks.go

```diff
@@ -139,6 +140,11 @@ type Hooks interface {
 	// If actor is not nil, the user was removed from the team by the actor.
 	UserHasLeftTeam(c *Context, teamMember *model.TeamMember, actor *model.User)
 
+	// HealthCheck is invoked periodically to assert the plugin's current health.
+	// If the health check fails more than the configured number of times, the process will be destroyed and restarted.
+	//
+	HealthCheck() error
```

Comment:
I can't think of a use case where restarting the plugin helps with the problem. Can someone please name one for me?

Reply:
The plugin maybe deadlocked, process running but not responsive.

Reply:
Could also be a rare crash where the issue only happens infrequently so restarting the plugin brings it back up for common case use

Reply:
> The plugin maybe deadlocked, process running but not responsive.

If the plugin can detect it's own deadlock, why shoun't it just `panic`?

My questions is: Why should we poll periodically if the plugin is unhealthy, instead of waiting for the plugin to send us this information? (Via e.g. a `Exit()` API method)

Reply:
Sorry for being persistent here, but I would like to remove this discussion before continuing with further review.

@levb @jwilander @mickmister I would love to hear you thoughts on this.

Reply:
If we were to support the Exit() API method, can we assert that it is being used, much less being used properly on the plugin side? I think adding more methods to be able to gather info on the current state of the plugin is right, but I don't think we should throw out the mechanism of monitoring it directly, from the outside.

I think the process check and regular RPC ping are the most useful, because it will work out of the box for all plugins. To me, the HealthCheck hook (and the Exit() API  method) are additional pieces that require the plugin author to follow an optional convention. That may change but it should probably remain backwards compatible.

Re-reading the context of your question, you may just be talking about the HealthCheck hook and not the other two mechanisms. If the plugin is written in a way that can tell us when it is crashing, that makes sense. But if it is unhealthy by some other metric or it is not panicking/crashing, asking it periodically seems to be a good way to find out about those issues, so it doesn't have to run its own interval to continuously run those checks.

Reply:
Another option for the plugin author would be to have the plugin run its own checks at the beginning and/or end of each hook/http call. Then the plugin could call `Exit()` whenever it saw fit, after a self-check determined it was unhealthy. Is this what you had in mind @hanzei?

I like this approach because the plugin author or admin will know the context of where in execution it is exiting, versus arbitrarily running a check when the server has asked it to. Both approaches have their own merits, and they certainly aren't mutually exclusive.

The HealthCheck hook has the potential to collect info on the plugin's status, without it necessarily being an exit condition. The hook would need to support a return type that signals that it does not need to be restarted. This could, in the future, integrate with external monitoring systems to support thorough diagnostic reporting for admins, including health of the Mattermost server itself.

Reply:
> Then the plugin could call `Exit()` whenever it saw fit, after a self-check determined it was unhealthy. Is this what you had in mind @hanzei?

As hooks are the only "entry point" for executing plugin code: Yes :wink:.

> The HealthCheck hook has the potential to collect info on the plugin's status, without it necessarily being an exit condition.

Maybe I misunderstood your proposal, but I thought the plugin would be restarted every time the `HealthCheck` returns an error. Or are you talking about just collecting data and not returning an error?

> The hook would need to support a return type that signals that it does not need to be restarted. 

I really like this idea. :+1: This would also close the discussion about when to restart the plugin and when not.


Do we agree that a `Exit` API method is a good idea? Maybe with the same return type that signals that it does not need to be restarted?

If so, do you want to include this in this PR or just open a ticket for this?

Reply:
I like the idea of using a push protocol to communicate to the Mattermost server that there is an issue, so I'm all for the `Exit()` method. If this method would support a non-restart field, then I suggest we name it something other than `Exit`. It's making me think that we should support three situations here:

- trivial error, no action necessary other than reporting the error
- fatal error, restart plugin
- fatal error, kill plugin without restarting

In any case, these messages should be reported to the admin somehow, which will be covered in a follow-up PR addressing UX issues around these monitoring/restart changes. I would say that supporting this API method in general is out of scope of this PR.

@levb, any thoughts on this?

Reply:
I really like the 3 conditions described, 2/5 I'd return an `error` and a separate `bool` to `DoNotRestart` (0/5 on the default behavior and therefore the naming).

Not sure I am a fan of push notifications and/or plugins managing their own restarts, in fact 1/5 predisposed against both. IMO ideally, like any production code, plugins should be free of panics. IMO pull is generally a more predictable way of reliably gathering data for centralized processing. 0/5 pull results in less code.

Reply:
> I really like the 3 conditions described, 2/5 I'd return an `error` and a separate `bool` to `DoNotRestart` (0/5 on the default behavior and therefore the naming).

But we need more then a `bool` to describe the three conditions. Or do I get your proposal wrong?

> IMO ideally, like any production code, plugins should be free of panics. 

I don't see them as panics. Think about e.g. a plugin requiring a specific config setting. I would check in `OnConfigurationChange` if this setting is pressed and otherwise `Exit()'. 

> IMO pull is generally a more predictable way of reliably gathering data for centralized processing.

Could you elaborate a bit?

Reply:
> But we need more then a bool to describe the three conditions. Or do I get your proposal wrong?

I agree, a single boolean would not allow for all cases to be handled. I think it's better to think of it as four conditions, where one condition is that no error occurred. So it would be something like this:
- no error: `return nil, nil`
- trivial error: `return err, nil` or `return err, Actions.LEAVE_RUNNING`
- error that warrants restart: `return err, Actions.KILL_AND_RESTART`
- error that should just kill plugin (like config error) `return err, Actions.KILL_WITHOUT_RESTART`

However, this implementation seems a bit awkward. We may not want the plugin author to be able to differentiate between `kill` and `kill and restart`, if we  want to avoid this level of granularity.

Reply:
> We may not want the plugin author to be able to differentiate between `kill and restart` and `kill`.

Why would we not want allow plugins to differ between these two? E.g. if a plugins deadlock it would call `kill and restart` and if I see a not supported config value it would call `kill'.

Reply:
@hanzei I agree, please disregard my comment about limiting the options. We would want to give options to solve those cases as you point out.

To account for all three error cases, I'm thinking we can use an enum, similar to the [constants](https://github.com/mattermost/mattermost-server/blob/387482f77b2239182100cb99252d390601fb1bd3/plugin/hooks.go#L17) that are used as ids to differentiate between the available hooks. An example for signaling a specific error case would be something like `PLUGIN_HEALTH_CHECK_IGNORE_ERROR`, `PLUGIN_HEALTH_CHECK_KILL`, and `PLUGIN_HEALTH_CHECK_RESTART`. What do you think of this method?

Reply:
:+1: for enums. I would like to use a typed one to make clear what the expected values are. Like described in https://stackoverflow.com/a/14426447. Are you fine with this?

To be fair: I have no good idea for naming. I would like to avoid `HEALTH_CHECK` as I want to use the same enum for an `Exit()` method later. Maybe something like `PLuginExitCommand`? 0/5

Reply:
One additional thought:

Why do we need a specific enum anyway? Why not go with two methods `Exit()` and `Restart()`. The error returned in the healthcheck hook would just be logged. And if a plugin needs to exit in some way, it can call the API methods.

I know this idea is a bit late in the discussion, but this design would make it much more clear, what the healtheck does and what not.

Reply:
At the point of writing, it's been established that a failed process check or RPC ping should warrant a plugin restart, but there is discussion in this thread on other cases where the plugin should be restarted. The two additional methods being discussed are:

- allow restarting of the plugin by the server, after retrieving an error from the HealthCheck hook (after process check and ping pass)
- allow the plugin author to choose when to restart or kill based on new plugin API methods `Exit()` and `Restart()` and the HealthCheck hook would not restart the plugin

Along with the idea of allowing the HealthCheck hook result to restart the plugin, there is also discussion on letting the returned error have a conditional signal to tell the server what action to take on an unhealthy plugin. The server could use the signal to kill and restart the plugin, use the signal to explicitly kill the plugin without restarting, or use the signal to say that the error is benign and shouldn't be used to kill or restart the plugin. Using API methods from the plugin side to restart instead of the HealthCheck hook would mean that we don't need these special flags.

I'm inclined to remove the HealthCheck hook from this PR and keep it as just the process check and ping for the purposes of purely introducing code for monitoring and restarting plugins. Adding any application-level options increases the scope of discussion.

@crspeller @jwilander Do you have thoughts on these points, or thoughts on what should be considered in the scope of this PR in regards to this discussion?

Reply:
> I'm inclined to remove the HealthCheck hook from this PR and keep it as just the process check and ping for the purposes of purely introducing code for monitoring and restarting plugins. Adding any application-level options increases the scope of discussion.

100% agree on this. Sorry for driving away the discussion from the original idea of this PR.